engineapi/backups.go

```diff
@@ -42,17 +41,36 @@ func getBackupCredentialEnv(backupTarget string, credential map[string]string) (
 		return envs, err
 	}
 	if backupType == types.BackupStoreTypeS3 {
-		if credential != nil && credential[types.AWSAccessKey] != "" && credential[types.AWSSecretKey] != "" {
-			envs = append(envs, fmt.Sprintf("%s=%s", types.AWSAccessKey, credential[types.AWSAccessKey]))
-			envs = append(envs, fmt.Sprintf("%s=%s", types.AWSSecretKey, credential[types.AWSSecretKey]))
+		if credential != nil {
+			if credential[types.AWSIAMRole] == "" {
+				if credential[types.AWSAccessKey] == "" && credential[types.AWSSecretKey] == "" {
+					return nil, fmt.Errorf("Could not backup to %s without setting credential", backupType)
+				}
+				if credential[types.AWSAccessKey] == "" && credential[types.AWSSecretKey] != "" {
+					return nil, fmt.Errorf("Could not backup to %s without setting credential access key", backupType)
+				}
+				if credential[types.AWSAccessKey] != "" && credential[types.AWSSecretKey] == "" {
+					return nil, fmt.Errorf("Could not backup to %s without setting credential secret access key", backupType)
+				}
```

Comment:
Example one how to flatten this whole if/else nesting.
You want to check your pre conditions first and exit, this way you don't need lots of inner if/else structures and can assume you are in a valid state.

I also added the missing keys, since this way the returned error is actually actionable for the user.

```go
if backupType != types.BackupStoreTypeS3 || credential == nil { 
    return env, nil
}

var missingKeys string;
if credential[types.AWSAccessKey] == "" {
    missingKeys = append(missingKeys, string(types.AWSAccessKey)
}

if credential[types.AWSSecretKey] == "" {
    missingKeys = append(missingKeys, string(types.AWSSecretKey)
}

areKeysRequired := credential[types.AWSIAMRole] == ""
if (areKeysRequired && len(missingKeys) > 0) {
    return nil, fmt.Errorf("Could not backup to %s secret is missing keys %v", backupType, missingKeys)
}

if len(missingKeys) == 0 {
    envs = append(envs, fmt.Sprintf("%s=%s", types.AWSAccessKey, credential[types.AWSAccessKey]))
    envs = append(envs, fmt.Sprintf("%s=%s", types.AWSSecretKey, credential[types.AWSSecretKey]))
}
```

Reply:
Thank u, it's excellent.